// computed-types/lib/DateType.js
"use strict";

exports.__esModule = true;
exports.default = exports.DateValidator = void 0;

var _Validator = _interopRequireDefault(require("./Validator"));

var _string = require("./string");

var _errors = require("./schema/errors");

var _number = require("./number");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class DateValidator extends _Validator.default {
  constructor(...args) {
    super(...args);
    this.gte = this.min;
    this.lte = this.max;
  }

  toISOString(...args) {
    return this.transform(val => val.toISOString(...args), _string.StringValidator);
  }

  getTime(...args) {
    return this.transform(val => val.getTime(...args), _number.NumberValidator);
  }

  min(min, error) {
    return this.test(val => val >= min, error || (val => new RangeError(`Expect date to be greater or equal than ${min} (actual: ${val})`)));
  }

  max(max, error) {
    return this.test(val => val <= max, error || (val => new RangeError(`Expect date to be lower or equal than ${max} (actual: ${val})`)));
  }

  gt(boundary, error) {
    return this.test(val => val > boundary, error || (val => new RangeError(`Expect date to be greater than ${boundary} (actual: ${val})`)));
  }

  lt(boundary, error) {
    return this.test(val => val < boundary, error || (val => new RangeError(`Expect date to be lower than ${boundary} (actual: ${val})`)));
  }

  between(min, max, error) {
    return this.test(val => val >= min && val <= max, error || (val => new RangeError(`Expect date to be between ${min} and ${max} (actual: ${val})`)));
  }

}

exports.DateValidator = DateValidator;
const DateType = new DateValidator(input => {
  if (!(input instanceof Date)) {
    throw (0, _errors.toError)(`Expect value to be instance of Date`);
  }

  return input;
}).proxy();
var _default = DateType;
exports.default = _default;
