// computed-types/lib/number.js
"use strict";

exports.__esModule = true;
exports.default = exports.NumberValidator = void 0;

var _Validator = _interopRequireDefault(require("./Validator"));

var _string = require("./string");

var _validations = require("./schema/validations");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class NumberValidator extends _Validator.default {
  constructor(...args) {
    super(...args);
    this.gte = this.min;
    this.lte = this.max;
  }

  float(error) {
    return this.test(val => !isNaN(val) && Number.isFinite(val), error || `Expect value to be a number`);
  }

  integer(error) {
    return this.test(val => Number.isInteger(val), error || `Expect value to be an integer`);
  }

  toExponential(...args) {
    return this.transform(val => val.toExponential(...args), _string.StringValidator);
  }

  toFixed(...args) {
    return this.transform(val => val.toFixed(...args), _string.StringValidator);
  }

  toLocaleString(...args) {
    return this.transform(val => val.toLocaleString(...args), _string.StringValidator);
  }

  toPrecision(...args) {
    return this.transform(val => val.toPrecision(...args), _string.StringValidator);
  }

  toString(...args) {
    return this.transform(val => val.toString(...args), _string.StringValidator);
  }

  min(min, error) {
    return this.test(val => val >= min, error || (val => new RangeError(`Expect value to be greater or equal than ${min} (actual: ${val})`)));
  }

  max(max, error) {
    return this.test(val => val <= max, error || (val => new RangeError(`Expect value to be lower or equal than ${max} (actual: ${val})`)));
  }

  gt(boundary, error) {
    return this.test(val => val > boundary, error || (val => new RangeError(`Expect value to be greater than ${boundary} (actual: ${val})`)));
  }

  lt(boundary, error) {
    return this.test(val => val < boundary, error || (val => new RangeError(`Expect value to be lower than ${boundary} (actual: ${val})`)));
  }

  between(min, max, error) {
    return this.test(val => val >= min && val <= max, error || (val => new RangeError(`Expect value to be between ${min} and ${max} (actual: ${val})`)));
  }

}

exports.NumberValidator = NumberValidator;
const number = new NumberValidator((0, _validations.type)('number')).proxy();
var _default = number;
exports.default = _default;
