// computed-types/lib/string.js
"use strict";

exports.__esModule = true;
exports.default = exports.StringValidator = void 0;

var _Validator = _interopRequireDefault(require("./Validator"));

var _validations = require("./schema/validations");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class StringValidator extends _Validator.default {
  toLowerCase() {
    return this.transform(str => str.toLowerCase());
  }

  toUpperCase() {
    return this.transform(str => str.toUpperCase());
  }

  toLocaleLowerCase(...input) {
    return this.transform(str => str.toLocaleLowerCase(...input));
  }

  toLocaleUpperCase(...input) {
    return this.transform(str => str.toLocaleUpperCase(...input));
  }

  normalize(...input) {
    return this.transform(str => str.normalize(...input));
  }

  trim() {
    return this.transform(str => str.trim());
  }

  truncate(length) {
    return this.transform(str => str.length > length ? `${str.substring(0, length - 1)}…` : str);
  }

  min(length, error) {
    return this.test(str => str.length >= length, error || (str => new RangeError(`Expect length to be minimum of ${length} characters (actual: ${str.length})`)));
  }

  max(length, error) {
    return this.test(str => str.length <= length, error || (str => new RangeError(`Expect length to be maximum of ${length} characters (actual: ${str.length})`)));
  }

  between(minLength, maxLength, error) {
    return this.test(str => str.length >= minLength && str.length <= maxLength, error || (str => new RangeError(`Expect length to be between ${minLength} and ${maxLength} characters (actual: ${str.length})`)));
  }

  regexp(exp, error) {
    return this.transform((0, _validations.regexp)(exp, error));
  }

}

exports.StringValidator = StringValidator;
const string = new StringValidator((0, _validations.type)('string')).proxy();
var _default = string;
exports.default = _default;
