// computed-types/lib/unknown.js
"use strict";

exports.__esModule = true;
exports.default = exports.UnknownValidator = void 0;

var _Validator = _interopRequireDefault(require("./Validator"));

var _errors = require("./schema/errors");

var _object = require("./object");

var _string = require("./string");

var _number = require("./number");

var _boolean = require("./boolean");

var _compiler = _interopRequireDefault(require("./schema/compiler"));

var _array = require("./array");

var _validations = require("./schema/validations");

var _DateType = require("./DateType");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const BOOL_MAP = {
  true: true,
  false: false,
  t: true,
  f: false,
  yes: true,
  no: false,
  y: true,
  n: false,
  1: true,
  0: false
};

class UnknownValidator extends _Validator.default {
  schema(schema, error) {
    return this.transform((0, _compiler.default)(schema, {
      error
    }), _Validator.default);
  }

  object(error) {
    return this.transform((0, _validations.type)('object', error), _object.ObjectValidator);
  }

  array(error) {
    return this.transform((0, _validations.array)(null, error), _array.ArrayValidator);
  }

  string(error) {
    return this.transform(input => {
      if (typeof input === 'string') {
        return input;
      }

      if (input == null || typeof input === 'object' && // eslint-disable-next-line @typescript-eslint/ban-types
      input.toString === Object.prototype.toString) {
        throw (0, _errors.toError)(error || `Expect value to be string`, input);
      }

      return String(input);
    }, _string.StringValidator);
  }

  number(error) {
    return this.transform(input => {
      if (typeof input === 'number') {
        return input;
      }

      const value = Number(input);

      if (isNaN(value) && input !== 'NaN') {
        throw (0, _errors.toError)(error || `Unknown number value`, input);
      }

      return value;
    }, _number.NumberValidator);
  }

  boolean(error) {
    return this.transform(input => {
      if (typeof input === 'boolean') {
        return input;
      }

      const key = String(input).trim().toLowerCase();
      const value = BOOL_MAP[key];

      if (value == null) {
        throw (0, _errors.toError)(error || `Unknown boolean value`, input);
      }

      return value;
    }, _boolean.BooleanValidator);
  }

  date(error) {
    return this.transform(input => {
      if (input instanceof Date) {
        return input;
      }

      if (typeof input === 'number' || typeof input === 'string') {
        const value = new Date(input);

        if (!isNaN(value.getTime())) {
          return value;
        }
      }

      throw (0, _errors.toError)(error || `Unknown date value`, input);
    }, _DateType.DateValidator);
  }

  enum(value, error) {
    return this.transform((0, _validations.enumValue)(value, error), _Validator.default);
  }

  record(key, // eslint-disable-line @typescript-eslint/no-explicit-any
  value, // eslint-disable-line @typescript-eslint/no-explicit-any
  error) {
    return this.transform((0, _validations.recordValue)(key, value, error), _Validator.default);
  }

}

exports.UnknownValidator = UnknownValidator;
const unknown = new UnknownValidator(input => input).proxy();
var _default = unknown;
exports.default = _default;
