// computed-types/lib/Validator.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _validations = require("./schema/validations");

var _utils = require("./schema/utils");

var _logic = require("./schema/logic");

class Validator {
  constructor(validator) {
    this.validator = void 0;
    this.validator = validator;
  }

  proxy() {
    return new Proxy(this.validator, {
      get: (target, propertyKey) => propertyKey in this ? this[propertyKey] : this.validator[propertyKey]
    });
  }

  equals(value, error) {
    return this.transform((0, _validations.equals)(value, error));
  }

  test(tester, error) {
    return this.transform((0, _validations.test)(tester, error));
  }

  transform(fn, constructor = this.constructor) {
    const {
      validator
    } = this;
    return new constructor((...args) => {
      const res = validator(...args);

      if (!(0, _utils.isPromiseLike)(res)) {
        return fn(res);
      }

      return res.then(ret => fn(ret));
    }).proxy();
  }

  construct(fn) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const Class = this.constructor;
    const {
      validator
    } = this;
    return new Class((...args) => validator(...fn(...args))).proxy();
  }

  optional(defaultValue) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const Class = this.constructor;
    const {
      validator
    } = this;
    return new Class((0, _logic.optional)(validator, defaultValue)).proxy();
  }

  strictOptional(defaultValue) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const Class = this.constructor;
    const {
      validator
    } = this;
    return new Class((0, _logic.strictOptional)(validator, defaultValue)).proxy();
  }

  destruct(error) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const Class = this.constructor;
    const {
      validator
    } = this;
    return new Class((0, _validations.destruct)(validator, error)).proxy();
  }

  error(err) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const Class = this.constructor;
    const {
      validator
    } = this;
    return new Class((0, _validations.error)(validator, err)).proxy();
  }

} // circular dependencies: import those after creating the validator class
// import { ObjectValidator } from './object';
// import { StringValidator } from './string';
// import { NumberValidator } from './number';
// import { BooleanValidator } from './boolean';


exports.default = Validator;
