// computed-types/lib/schema/errors.js
"use strict";

exports.__esModule = true;
exports.ValidationError = void 0;
exports.createValidationError = createValidationError;
exports.toError = toError;
exports.toPathErrors = toPathErrors;

class ValidationError extends TypeError {
  constructor(message, errors) {
    super(message);
    this.errors = void 0;
    this.path = void 0;
    this.errors = errors; // Extending error in TypeScript:
    // https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work

    Object.setPrototypeOf(this, ValidationError.prototype);
  }

  toJSON() {
    var _this$errors;

    return {
      message: this.message,
      errors: (_this$errors = this.errors) == null ? void 0 : _this$errors.map(({
        path,
        error
      }) => {
        var _ValidationError$prot;

        return {
          path,
          error: ((_ValidationError$prot = ValidationError.prototype.toJSON) == null ? void 0 : _ValidationError$prot.apply(error)) || error
        };
      })
    };
  }

}

exports.ValidationError = ValidationError;
ValidationError.prototype.name = 'ValidationError';

function toError(error, ...args) {
  if (typeof error === 'string') {
    return new ValidationError(error);
  }

  if (typeof error === 'function') {
    return toError(error(...args));
  }

  return error;
}

function toPathErrors(errorLike, path) {
  var _error$errors;

  const error = toError(errorLike);

  if (Array.isArray(error.path)) {
    path = error.path;
  }

  if ((_error$errors = error.errors) != null && _error$errors.length) {
    return [].concat(...error.errors.map(item => toPathErrors(item.error, [...path, ...item.path])));
  }

  return [{
    error,
    path
  }];
}

function createValidationError(errors, error, ...args) {
  if (!error) {
    if (errors[0]) {
      const {
        path,
        error: err
      } = errors[0];
      const message = String(err && err.message || err);
      error = path ? `${path.join('.')}: ${message}` : message;
    } else {
      error = 'Unknown Validation Error';
    }
  }

  const err = toError(error, ...args);
  err.errors = errors;
  return err;
}
