// computed-types/lib/schema/utils.js
"use strict";

exports.__esModule = true;
exports.deepConcat = deepConcat;
exports.isPrimitive = isPrimitive;
exports.isPromiseLike = isPromiseLike;
exports.typeCheck = typeCheck;

var _errors = require("./errors");

// export type RequiredKeys<T> = {
//   [k in keyof T]-?: undefined extends T[k] ? never : k;
// }[keyof T];
// export type OptionalKeys<T> = Exclude<keyof T, RequiredKeys<T>>;
// exported functions
function typeCheck(ok) {
  return ok;
}

function isPromiseLike(value) {
  return !!value && typeof value.then === 'function';
}

function isPrimitive(value) {
  return typeof value !== 'object' && typeof value !== 'function' || value === null;
}

function deepConcat(...values) {
  if (values.length < 2) {
    return values[0];
  }

  values = values.filter(value => value !== undefined);

  if (values.length < 2) {
    return values[0];
  }

  let base = values[0];

  if (typeof base !== 'object' || base === null) {
    for (let i = 1; i < values.length; i += 1) {
      if (values[i] !== base) {
        throw new _errors.ValidationError(`Type mismatch on validation concat`);
      }
    }

    return base;
  }

  const keys = {};

  for (let i = 0; i < values.length; i += 1) {
    const value = values[i];

    if (typeof value !== 'object' || value === null) {
      throw new _errors.ValidationError(`Type mismatch on validation concat`);
    }

    for (const key in value) {
      if (!Object.prototype.hasOwnProperty.call(value, key)) continue;
      const keyVal = value[key];
      if (keyVal === undefined) continue;

      if (!keys[key]) {
        keys[key] = [];
      }

      keys[key].push(keyVal);
    }
  }

  if (Array.isArray(base)) {
    base = [];
  } else {
    base = {};
  }

  for (const key in keys) {
    base[key] = deepConcat(...keys[key]);
  }

  return base;
}
