// computed-types/lib/schema/validations.js
"use strict";

exports.__esModule = true;
exports.array = array;
exports.destruct = destruct;
exports.enumValue = enumValue;
exports.equals = equals;
exports.error = error;
exports.recordValue = recordValue;
exports.regexp = regexp;
exports.test = test;
exports.type = type;

var _errors = require("./errors");

var _utils = require("./utils");

function type(type, error) {
  return (...args) => {
    if (typeof args[0] !== type || args[0] === null) {
      throw (0, _errors.toError)(error || `Expect value to be "${type}"`, ...args);
    }

    return args[0];
  };
}

function equals(expected, error) {
  return (...args) => {
    if (args[0] !== expected) {
      throw (0, _errors.toError)(error || `Expect value to equal "${expected}"`, ...args);
    }

    return args[0];
  };
}

function test(tester, error) {
  return (...args) => {
    if (!tester(...args)) {
      throw (0, _errors.toError)(error || `Validation test failed`, ...args);
    }

    return args[0];
  };
}

function destruct(validator, error) {
  return (...args) => {
    try {
      const res = validator(...args);

      if (!(0, _utils.isPromiseLike)(res)) {
        return [null, res];
      }

      return res.then(ret => [null, ret], err => [error ? (0, _errors.toError)(error, ...args) : err]);
    } catch (err) {
      return [error ? (0, _errors.toError)(error, ...args) : err];
    }
  };
}

function error(validator, err) {
  return (...args) => {
    try {
      const res = validator(...args);

      if (!(0, _utils.isPromiseLike)(res)) {
        return res;
      }

      return res.then(null, () => {
        throw (0, _errors.toError)(err, ...args);
      });
    } catch (e) {
      throw (0, _errors.toError)(err, ...args);
    }
  };
}

function regexp(exp, error) {
  if (!(exp instanceof RegExp)) {
    exp = new RegExp(exp);
  }

  return (...args) => {
    if (!exp.test(args[0])) {
      throw (0, _errors.toError)(error || `Invalid string format (expected: ${exp})`, ...args);
    }

    return String(args[0]);
  };
}

function array(length = null, error) {
  const isArray = (...args) => {
    if (!Array.isArray(args[0])) {
      throw (0, _errors.toError)(error || `Expecting value to be an array`, ...args);
    }

    return args[0];
  };

  if (length === null) {
    return isArray;
  }

  return (...args) => {
    const arr = isArray(...args);

    if (arr.length !== length) {
      throw (0, _errors.toError)(error || `Expected array length ${length} (given: ${arr.length})`, ...args);
    }

    return arr;
  };
}

function enumValue(value, error) {
  const values = new Set(Object.keys(value).filter(key => isNaN(Number(key))).map(key => value[key]));
  return (...args) => {
    if (!values.has(args[0])) {
      throw (0, _errors.toError)(error || 'Unknown enum value', ...args);
    }

    return args[0];
  };
}

function recordValue(key, value, error) {
  return (...args) => {
    const [input] = args;

    if (!input || typeof input !== 'object') {
      throw (0, _errors.toError)(error || 'Expected non-null object', ...args);
    }

    const obj = {};

    for (const k in input) {
      try {
        obj[key(k)] = value(input[k]);
      } catch (error) {
        throw (0, _errors.createValidationError)([{
          path: [k],
          error: error
        }], null, ...args);
      }
    }

    return obj;
  };
}
